File: app/src/main/java/com/xnote/app/domain/rules/NotebookRules.kt
package com.xnote.app.domain.rules

import com.xnote.app.domain.model.Note
import com.xnote.app.domain.model.Notebook

// -- Type Definitions

data class NotebookDeletionPatch(
    val noteId: String,
    val originalNotebookName: String,
    val deletedAtEpochMs: Long,
)

// -- Functions

fun patchesForDeletedNotebook(
    notebook: Notebook,
    notesInNotebook: List<Note>,
    nowMs: Long,
): List<NotebookDeletionPatch> {
    return notesInNotebook.map { note ->
        NotebookDeletionPatch(
            noteId = note.id,
            originalNotebookName = notebook.name,
            deletedAtEpochMs = note.deletedAtEpochMs ?: nowMs,
        )
    }
}

fun notebookIdAfterRestore(note: Note, notebookExists: (String) -> Boolean): String? {
    val notebookId = note.notebookId ?: return null
    return if (notebookExists(notebookId)) notebookId else null
}

File: flutter/lib/domain/rules/notebook_rules.dart
import '../model/note.dart';
import '../model/notebook.dart';

// -- Type Definitions

final class NotebookDeletionPatch {
  const NotebookDeletionPatch({
    required this.noteId,
    required this.originalNotebookName,
    required this.deletedAtEpochMilliseconds,
  });

  final String noteId;
  final String originalNotebookName;
  final int deletedAtEpochMilliseconds;

  // -- Functions

  @override
  bool operator ==(Object other) {
    return other is NotebookDeletionPatch &&
        other.noteId == noteId &&
        other.originalNotebookName == originalNotebookName &&
        other.deletedAtEpochMilliseconds == deletedAtEpochMilliseconds;
  }

  @override
  int get hashCode => Object.hash(
        noteId,
        originalNotebookName,
        deletedAtEpochMilliseconds,
      );
}

// -- Functions

List<NotebookDeletionPatch> patchesForDeletedNotebook({
  required Notebook notebook,
  required List<Note> notesInNotebook,
  required int nowEpochMilliseconds,
}) {
  return List<NotebookDeletionPatch>.unmodifiable(
    notesInNotebook.map(
      (note) => NotebookDeletionPatch(
        noteId: note.id,
        originalNotebookName: notebook.name,
        deletedAtEpochMilliseconds:
            note.deletedAtEpochMilliseconds ?? nowEpochMilliseconds,
      ),
    ),
  );
}

String? notebookIdAfterRestore(
  Note note,
  bool Function(String notebookId) notebookExists,
) {
  final notebookId = note.notebookId;
  return notebookId != null && notebookExists(notebookId) ? notebookId : null;
}
